### ISSUE TITLE
MSAL Error: AADSTS9001023

### ISSUE THREAD

**Issue by v-zhangw** [AzureAD/microsoft-authentication-library-for-java#382]
After moving from ADAL 1.6.6 to MSAL 1.91, an error was encountered when trying to connect to an Azure Data Explorer endpoint.

Error reads as follows:
AADSTS9001023: The grant type is not supported over the /common or /consumers endpoints. Please use the /organizations or tenant-specific endpoint.

Could I get some information on what this error means and how it could be resolved?

**Comment by SomkaPe** [comment]
@v-zhangw  Error suggests to use authority with "/organizations" or "/tenantId" suffix instead of "common"

**Comment by cosh** [comment]
+1

**Comment by v-zhangw** [comment]
Hello,
The issue turned out to be due to the migration from ADAL to MSAL causing /common to be no longer acceptable for the customer's endpoint.

**Comment by cheenamalhotra** [comment]
This issue should be reopened.

The same works with MSAL.NET. If .NET can provide access token, why not Java?
Also Interactive authentication modes work fine - only username/password fails with "/common".

Also reported to Azure SDK for Java: https://github.com/Azure/azure-sdk-for-java/issues/23146 as Azure Idenitity's UsernamePasswordCredential is also impacted.

**Comment by sangonzal** [comment]
Hi @cheenamalhotra can you point me to anywhere that shows this being supported in MSAL.NET? 

As far as I know, this AAD service behavior, regardless of the client that you use. If you take a look at the [MSAL.NET wiki page](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/Username-Password-Authentication#authority-implications) on UsernamePassword flow, you'll see that they also don't say that they don't support this flow for /common or /consumer authorities. 

To summarize, if you would like to use Username passoword flow, you either have to use /organization or /tenantid endpoints (neither common nor consumers will work). This is not a Java specific thing, instead it's the default behavior by the AAD service.

**Comment by cheenamalhotra** [comment]
@sangonzal 

I retested and I think I might have confused with AAD Interactive/MFA for the same user I tested AAD Password Authentication with, as MFA Authentication works in this case. I can confirm I reproduce same error with MSAL.NET.

Sorry for the confusion :(

The thing is we don't own these endpoints, it comes from Azure Data Explorer, we don't modify and pass on as-is to MSAL. Would you recommend this to Azure Data Explorer to fix this?

**Comment by sangonzal** [comment]
Chatted on teams, @cheenamalhotra has an email thread with the Azure Data Explorer team, where we will continue this conversation.

**Comment by j2-z** [comment]
@cheenamalhotra  @sangonzal Could I know how is this issue going on? I tried with the latest MSSQL JDBC and MSAL4J lib. The issue still presents.

```pom
  <dependencies>
    <dependency>
      <groupId>com.microsoft.sqlserver</groupId>
      <artifactId>mssql-jdbc</artifactId>
      <version>9.4.0.jre8</version>
    </dependency>
    <dependency>
      <groupId>com.microsoft.azure</groupId>
      <artifactId>msal4j</artifactId>
      <version>1.11.0</version>
    </dependency>
  </dependencies>
```

```
Caused by: java.lang.RuntimeException: AADSTS9001023: The grant type is not supported over the /common or /consumers endpoints. Please use the /organizations or tenant-specific endpoint.
Timestamp: 2021-08-18 06:03:03Z
	at com.microsoft.sqlserver.jdbc.SQLServerMSAL4JUtils.getCorrectedException(SQLServerMSAL4JUtils.java:218)
	... 21 more
```

As @cheenamalhotra  mentioned in https://github.com/Azure/azure-sdk-for-java/issues/23146, our application is using MSSQL JDBC interface and it's not possible to change the authentication endpoint. What should we do to solve this problem? Is it an issue in MSAL4J lib?  Thanks.

**Comment by j2-z** [comment]
I noticed that when MSSQL JDBC connects to an ADX server, the STSURL it received is `https://login.microsoftonline.com/common`
I wonder if it returns `https://login.microsoftonline.com/{tenantID}` then it might be able to work.
Is there any way to configure in ADX or Azure side to make ADX returns `https://login.microsoftonline.com/{tenantID} as STSURL ? `
```
Aug 18, 2021 4:06:40 PM com.microsoft.sqlserver.jdbc.SQLServerConnection processFedAuthInfo
FINE: ConnectionID:1 ClientConnectionId: xxxxx FedAuthInfoData: https://xxx.eastus.kusto.windows.net/
Aug 18, 2021 4:06:40 PM com.microsoft.sqlserver.jdbc.SQLServerConnection processFedAuthInfo
FINE: ConnectionID:1 ClientConnectionId: xxxxx FedAuthInfoOpt: ID=1, DataLen=80, Offset=104
Aug 18, 2021 4:06:40 PM com.microsoft.sqlserver.jdbc.SQLServerConnection processFedAuthInfo
FINE: ConnectionID:1 ClientConnectionId: xxxxx FedAuthInfoData: https://login.microsoftonline.com/common
```

**Comment by yogilad** [comment]
Hi,

We recently added a workaround to control the authority to authenticate to.
You can change authority , via SQL connection string _Language_ or _application_ fields (depends where and how each available on specific client – ADO.NET, ODBC, JDBC, etc.), for example:
```
Language=English@AadAuthority:YYYY
application=myappname@AadAuthority:YYYY
```

You can attempt this with the `organizations` keyword or with your specific tenant Id, and see if it helps.
If it does the change can later be applied as a default on the cluster.
